libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/model/Store.kt

```diff
@@ -0,0 +1,84 @@
+package com.mapbox.navigation.dropin.model
+
+import android.location.Location
+import com.mapbox.navigation.dropin.component.audioguidance.AudioGuidanceState
+import com.mapbox.navigation.dropin.component.camera.CameraState
+import com.mapbox.navigation.dropin.component.navigation.NavigationState
+import com.mapbox.navigation.dropin.component.routefetch.RoutesState
+import com.mapbox.navigation.dropin.component.tripsession.TripSessionStarterState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import java.util.concurrent.ConcurrentLinkedQueue
+
+data class State(
+    val destination: Destination? = null,
+    val location: Location? = null,
+    val navigation: NavigationState = NavigationState.FreeDrive,
+    val camera: CameraState = CameraState(),
+    val audio: AudioGuidanceState = AudioGuidanceState(),
+    val routes: RoutesState = RoutesState.Empty,
+    val tripSession: TripSessionStarterState = TripSessionStarterState()
+)
+
+interface Action
+
+fun interface Reducer {
+    fun process(state: State, action: Action): State
+}
+
+internal open class Store(
+    private val coroutineScope: CoroutineScope =
+        CoroutineScope(SupervisorJob() + Dispatchers.Main)
+) {
+    protected val _state = MutableStateFlow(State())
+    val state: StateFlow<State> = _state.asStateFlow()
+
+    private val actionBus = MutableSharedFlow<Action>()
+    private val reducers = ConcurrentLinkedQueue<Reducer>()
+
+    init {
+        coroutineScope.launch {
+            actionBus.collect { action ->
+                reducers.forEach { reducer ->
+                    if (isActive) {
+                        _state.value = reducer.process(_state.value, action)
```

Comment:
Adding a concern that there may be too much noise here.

For example, the `State.location` happens every second. In some cars it happens 10 times a second. Decoupling some states may be necessary at some point.

But we are notifying all reducers of state changes that only 1 of them care about. If we scale to have 50 reducers, and locations happening 10 times a second.. Do you have concerns about this loop becoming wasteful?

Reply:
In FLUX every reducer receives every dispatched action. 

Of course we can optimize this, but I'd rather wait until we identify that this is an actual bottleneck. 

One potential improvement/optimization could be introduction of a convenience method that would allow reducer registration with predicate.

```kotlin 
fun register(predicate: (Action) -> Boolean, reducer: Reducer): Registration {
    val r = Reducer { state, action ->
        if (predicate(action)) reducer.process(state, action)
        else state
    }
    reducers.add(r)
    return Registration {
        reducers.remove(r)
    }
}

fun interface Registration {
    fun unregister()
}
```